### PR TITLE
engine/metrics: populate closed_at field in alert_metrics

### PR DESCRIPTION
**Pull request by dctalbot**
- [x] Identified the issue which this PR solves.
- [x] Read the [**CONTRIBUTING**](https://github.com/target/goalert/blob/master/CONTRIBUTING.md) document.
- [x] Code builds clean without any errors or warnings.
- [x] Added appropriate tests for any new functionality.
- [x] All new and existing tests passed.
- [x] Added comments in the code, where necessary.
- [x] Ran `make check` to catch common errors. Fixed any that came up.

**Description:**
This PR populates a `closed_at` column in `alert_metrics` that will help with the implementation of #2122 

It also refactors the operation to work from oldest to newest, maintaining a `(timestamp, id)` cursor in the engine's state. To regenerate `alert_metrics` e.g. to catch any that were missed, one only needs to reset the engine state.

**How to validate:**
1. `make regendb` and `make start`
2. For each engine cycle, observe `alert_metrics` being populated (500 at a time) and the `engine_processing_version.state` value incrementing in a forward fashion
3. Wait for step 2 to finish, and observe that count(alert_metrics) == count(alerts where status = 'closed')
4. Create and close an alert
5. Wait two minutes before observing the new row in `alert_metrics` and the new `engine_processing_version.state` value

Optionally:

1. Delete some number of rows from `alert_metrics`
2. Clear `engine_processing_version.state` and observe the deleted rows regenerate

### CODE REVIEW

File: engine/metricsmanager/update.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 )
 
 type State struct {
-	V1 struct {
+	V2 struct {
```
Comment (mastercactapus) [inline engine/metricsmanager/update.go:14]: The struct/meaning hasn't changed, no need to increment the version

File: migrate/migrations/20220317153438-alert-metrics-add-closed-at.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+-- +migrate Up
+
+UPDATE engine_processing_versions
+SET "version" = 2
+WHERE type_id = 'metrics';
+
+ALTER TABLE alert_metrics ADD COLUMN closed_at TIMESTAMP WITH TIME ZONE;
```
Comment (mastercactapus) [inline migrate/migrations/20220317153438-alert-metrics-add-closed-at.sql:7]: We should make this not-null and truncate the table so we recalculate existing data

**Review by mastercactapus — CHANGES_REQUESTED** [review]
Were we going to change this to do a sliding time window instead of going by alert ID?

**Review by m17ch — DISMISSED** [review]
lgtm 👍

**Review by m17ch — APPROVED** [review]

**Review by mastercactapus — APPROVED** [review]

**Pull request merged**